src/utils/saml.ts

```diff
@@ -469,6 +470,272 @@ export const getAuthorizeRequestTamperer = (
 //  Validate response
 //
 
+const utcStringToDate = (value: string, tag: string): Date => {
+  const maybeDate =
+    value.length === 24
+      ? UTCISODateFromString.decode(value)
+      : UTCISODateFromString.decode(value.replace("Z", ".000Z"));
+  if (isLeft(maybeDate)) {
+    throw new Error(`${tag} must be an UTCISO format date string`);
+  }
+  return maybeDate.value;
+};
+
+const validateIssuer = (fatherElement: Element) => {
+  const IssuerElement = fatherElement
+    .getElementsByTagNameNS(SAML_NAMESPACE.ASSERTION, "Issuer")
+    .item(0);
+  if (
+    !IssuerElement ||
+    !IssuerElement.textContent ||
+    IssuerElement.textContent === ""
+  ) {
+    throw new Error("Issuer element must be present and not empty");
+  }
+  // TODO: Must validate that Issuer value is equal to EntityID IDP (29)(69)
+  /*if (IssuerElement.textContent !== samlConfig.idpIssuer) {
+    throw new Error(`"Invalid Issuer value: "${samlConfig.idpIssuer}`);
+  }*/
+  const IssuerFormatValue = IssuerElement.getAttribute("Format");
+  if (!IssuerFormatValue) {
+    throw new Error("Format attribute of Issuer element must be present");
+  }
+  if (
+    IssuerFormatValue !== "urn:oasis:names:tc:SAML:2.0:nameid-format:entity"
+  ) {
+    throw new Error("Format attribute of Issuer element is invalid");
+  }
+};
+
+const mainAttributeValidation = (requestOrAssertion: Element) => {
+  const ID = requestOrAssertion.getAttribute("ID");
+  if (!ID && ID === "") {
+    throw new Error("Assertion must contain a non empty ID");
+  }
+  const Version = requestOrAssertion.getAttribute("Version");
+  if (Version !== "2.0") {
+    throw new Error("Version version must be 2.0");
+  }
+  const IssueInstant = requestOrAssertion.getAttribute("IssueInstant");
+  if (!IssueInstant) {
+    throw new Error("Assertion must contain a non empty IssueInstant");
+  }
+  const IssueInstantDate = utcStringToDate(IssueInstant, "IssueInstant");
+  if (IssueInstantDate.getTime() > Date.now()) {
+    throw new Error("IssueInstant must be in the past");
+  }
+  return IssueInstantDate;
+};
+
+const isEmptyNode = (element: Element): boolean => {
+  if (element.childNodes.length > 1) {
+    return false;
+  } else if (
+    element.firstChild &&
+    element.firstChild.nodeType === element.ELEMENT_NODE
+  ) {
+    return false;
+  } else if (
+    element.textContent &&
+    element.textContent
+      ?.split("\r") // Exclude return character
```

Comment:
can we use something like `str.replace(/[\r\n\ ]+/g, "")` ?

Reply:
fixed